RecoTracker/TrackProducer/test/RefitWithSim.py

```diff
@@ -0,0 +1,186 @@
+
+
+### Track refitter specific stuff
+import RecoTracker.TrackProducer.TrackRefitter_cfi
+import CommonTools.RecoAlgos.recoTrackRefSelector_cfi
+process.mytkselector = CommonTools.RecoAlgos.recoTrackRefSelector_cfi.recoTrackRefSelector.clone()
+process.mytkselector.quality = ['highPurity']
+process.mytkselector.min3DLayer = 2
+process.mytkselector.ptMin = 0.5
+process.mytkselector.tip = 1.0
+
+process.FinalTracks = RecoTracker.TrackProducer.TrackRefitter_cfi.TrackRefitter.clone()
+process.FinalTracks.src= 'mytkselector'
+process.FinalTracks.NavigationSchool = ''
+process.FinalTracks.Fitter = 'FlexibleKFFittingSmoother'
+
+
+process.myRefittedTracks = RecoTracker.TrackProducer.TrackRefitter_cfi.TrackRefitter.clone()
+process.myRefittedTracks.src= 'mytkselector'  # or FinalTracks
+process.myRefittedTracks.NavigationSchool = ''
+process.myRefittedTracks.Fitter = 'FlexibleKFFittingSmoother'
+process.myRefittedTracks.TTRHBuilder = 'Fake'
+
+### and an analyzer
+process.trajCout = cms.EDAnalyzer('TrajectoryAnalyzer',
+   trajectoryInput=cms.InputTag('myRefittedTracks')
+)
+
+
+process.fakeCPEFiller = cms.EDFilter('FakeCPEFiller',
+    associatePixel = cms.bool(True), associateStrip = cms.bool(True), associateRecoTracks = cms.bool(False), 
+    ROUList = cms.vstring(
+    'TrackerHitsPixelBarrelLowTof',
+    'TrackerHitsPixelBarrelHighTof',
+    'TrackerHitsPixelEndcapLowTof',
+    'TrackerHitsPixelEndcapHighTof',
+    'TrackerHitsTIBLowTof', 
+    'TrackerHitsTIBHighTof', 
+    'TrackerHitsTIDLowTof', 
+    'TrackerHitsTIDHighTof', 
+    'TrackerHitsTOBLowTof', 
+    'TrackerHitsTOBHighTof', 
+    'TrackerHitsTECLowTof', 
+    'TrackerHitsTECHighTof'        
+                          ),  
+
+    stripSimLinkSrc = cms.InputTag("simSiStripDigis"),   
+    pixelSimLinkSrc = cms.InputTag("simSiPixelDigis")
+
+
+
+)
+
+
+process.myrefit = cms.Path(process.mytkselector+process.FinalTracks+process.fakeCPEFiller+process.myRefittedTracks+process.trajCout)
+
+
+# Schedule definition
+process.schedule = cms.Schedule(process.raw2digi_step,process.reconstruction_step,process.myrefit)
+from PhysicsTools.PatAlgos.tools.helpers import associatePatAlgosToolsTask
+associatePatAlgosToolsTask(process)
+
+#Setup FWK for multithreaded
+process.options.numberOfThreads=cms.untracked.uint32(1)
+process.options.numberOfStreams=cms.untracked.uint32(1)
+
+# customisation of the process.
+
+# Automatic addition of the customisation function from SimGeneral.MixingModule.fullMixCustomize_cff
+from SimGeneral.MixingModule.fullMixCustomize_cff import setCrossingFrameOn 
+
+#call to customisation function setCrossingFrameOn imported from SimGeneral.MixingModule.fullMixCustomize_cff
+process = setCrossingFrameOn(process)
+
+# End of customisation functions
+#do not add changes to your config after this point (unless you know what you are doing)
+from FWCore.ParameterSet.Utilities import convertToUnscheduled
+process=convertToUnscheduled(process)
```

Comment:
Actually, maybe it is better to remove the call to `convertToUnscheduled` altogether because the order of modules is important (because of the hidden dependence on how the fake CPE is filled).

Reply:
that's why I used a filter....

Reply:
Is reco supposed to support "scheduled" also in future? 
Here I am starting from RAW

Reply:
It doesn't really matter. Currently reco is (still) configured as scheduled, and the call to `convertToUnscheduled()` transforms it to unscheduled (=moves all EDProducers from Path to Task so they are run "on demand"). So, if this configuration works now (as unscheduled), it will also work without the call to `convertToUnscheduled()` and with reco (eventually) configured as unscheduled.

What you would gain here by leaving `convertToUnscheduled()` out is that you would be in full control of what happens, i.e. modules that have only a single instance in a Path will be always run, and in the order dictated by the Path, regardless of their type.

Reply:
Ok, in any case we need to better define what are the "best practices" we teach people.
I do not think everybody can trim his/her configuration at this level of detail.
(Actually it is my opinion that all this boilerplate in the cfg should go, and cfg should contain just the minimum needed by the user to say "run this part of standard config and btw also this other stuff of mine")

Reply:
> I do not think everybody can trim his/her configuration at this level of detail.

Certainly. On the other hand, people should not be communicating between modules and ESProducts this way, so in general there should be no need to do anything special.

> (Actually it is my opinion that all this boilerplate in the cfg should go, and cfg should contain just  the minimum needed by the user to say "run this part of standard config and btw also this other stuff of mine")

I agree, that would make life easier. My understanding is that achieving that would require cleaning up bunch of cff(/cfi?) configuration files.